### PR TITLE
Refactor key transformations

### PR DESCRIPTION
**Pull request by Aryk**
* Remove dependency on camel_snake_keys in favor of implementation
   already in activesupport
 * All specs pass
 * Move camel and snake functionality into their own utils folder so
   it can be reused throughout the gem
 * Minimize core extending by activesupport. As a gem, it should minimize side-effects in the users main application

### CODE REVIEW

**Comment by Aryk** [comment]
@buermann ping

**Comment by buermann** [comment]
Many pardons, I've been away. Will take a look at these today.

**Comment by buermann** [comment]
"Remove dependency on camel_snake_keys in favor of implementation
already in activesupport"

That doesn't seem to be what's going on here so much as re-implementing much of camel_snake_keys (which uses ActiveSupport's inflectors just the same) in introspective_grape. If the desire is to reduce the imports from ActiveSupport that can be done in the CamelSnakeKeys (which I think is already pretty minimal...) and lib/introspective_grape/camel_snake.rb, where I agree it should just be importing the necessary dependencies.

**Comment by buermann** [comment]
There's also a little edge case (the test coverage for it is in camel_snake_keys) I've run into in production where if you're camelizing a hash and it contains both the snaked and camelized key the (I think) proper behavior is to always give preference to the snaked key/value that would be transformed and should be presumed to be the intended value, but with this it's prioritized on reverse order in the hash:

irb(main):013:0* IntrospectiveGrape::Utils::KeyTransformations.camel_keys( { fooBar: 1, foo_bar: 2 } )
=> {"fooBar"=>2}
irb(main):014:0> IntrospectiveGrape::Utils::KeyTransformations.camel_keys( { foo_bar: 2, fooBar: 1 } )
=> {"fooBar"=>1}

And vice versa for snaking.

File: lib/introspective_grape/utils/json_expander.rb [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+    module JsonExpander
+      extend self
+
+      # Expands an object with its #to_json method if it's not a primitive or if the first element is a Grape::Entity.
+      def expand(object)
+        if (object.respond_to?(:to_json) &&
+          [String, Symbol, Hash, Array, NilClass, TrueClass, FalseClass, Numeric].all? { |x| !object.is_a?(x) }) ||
+          # Force arrays of Grape::Entities into their hash representations before camelizing
+          (object.is_a?(Array) && defined?(Grape::Entity) && object.first.is_a?(Grape::Entity))
+
```
Comment (buermann) [inline lib/introspective_grape/utils/json_expander.rb:12]: I remember the point of the increasingly complicated logic here was to not do `JSON.parse(object.to_json)` if the object was already a hash, or an array of hashes. Maybe something like

```
o = object.kind_of?(Array) ? object.first : object 
if o.respond_to?(:to_json) && !o.respond_to?(:with_camel_keys)
  JSON.parse(object.to_json)
else
  object
end
```

Would be simpler? I'm not a huge fan of my kludge for the array situation here expecting the array elements to all be the same class as the first and only dealing with the situation for Grape::Entity, since it would't work if the gem were to someday support other facade classes, but the assumption of like-elements seems like the lesser of many evils.

Comment (Aryk) [reply lib/introspective_grape/utils/json_expander.rb:12]: > I remember the point of the increasingly complicated logic here was to not do JSON.parse(object.to_json) if the object was already a hash, or an array of hashes. Maybe something like

So if that's the case, shouldn't we just directly check for that so

```ruby
(o.is_a?(Array) && o.first.is_a?(Hash)) || o.is_a?(Hash)
```

I'm not sure what the logic was, so I made a best guess, but based on what you are telling me, we should write the above.

What if the object was a primitive, like a [String, Symbol, Hash, Array, NilClass, TrueClass, FalseClass, Numeric], then we *should* do `JSON.parse(object.to_json)`? If so, why?

**Review by Aryk — COMMENTED** [review]
> "Remove dependency on camel_snake_keys in favor of implementation
already in activesupport"

> That doesn't seem to be what's going on here so much as re-implementing much of camel_snake_keys (which uses ActiveSupport's inflectors just the same) in introspective_grape. If the desire is to reduce the imports from ActiveSupport that can be done in the CamelSnakeKeys (which I think is already pretty minimal...) and lib/introspective_grape/camel_snake.rb, where I agree it should just be importing the necessary dependencies.

By, `CamelSnakeKeys` are you referring to the gem? If so, why exactly should we keep the "camel_snake_keys" dependency if *anyway* we have to keep ActiveSupport?

My point wasn't to remove ActiveSupport because that would be hard to do, but to reduce monkeypatching core classes, which my solution minimizes (along with removing a gem dependency). camel_snake_keys [injects](https://github.com/buermann/camel_snake_keys/blob/master/lib/camel_snake_keys.rb#L52) into Enumerable. What if a person has those methods defined in their app and you were changing the implementation of them?

**Comment by Aryk** [comment]
> There's also a little edge case (the test coverage for it is in camel_snake_keys) I've run into in production where if you're camelizing a hash and it contains both the snaked and camelized key the (I think) proper behavior is to always give preference to the snaked key/value that would be transformed and should be presumed to be the intended value, but with this it's prioritized on reverse order in the hash:

> irb(main):013:0* IntrospectiveGrape::Utils::KeyTransformations.camel_keys( { fooBar: 1, foo_bar: 2 } )
> => {"fooBar"=>2}
irb(main):014:0> IntrospectiveGrape::Utils::KeyTransformations.camel_keys( { foo_bar: 2, fooBar: 1 } )
=> {"fooBar"=>1}

> And vice versa for snaking.

I suppose we can throw in a test case in this gem to make sure things pass. Ultimately one could argue both ways I think, but ultimately matters more about the application of camel or snake casing the keys. For example, in Ruby there is "merge" and "reverse_merge", which one you use depends on the behavior you want. I'm fine to keep the behavior as it was before.

Assuming you're ok with the other changes, I can add a test case and ensure it does not overwrite, just doesn't really make sense to do it if you don't intend to merge. :)

**Comment by buermann** [comment]
"to reduce monkeypatching core classes, which my solution minimizes (along with removing a gem dependency)"

I get the first argument and not the second. Regarding the first I'd be happy to shove camel_snake_keys' monkey patches into a separate explicitly required module and change it to refine Enumerable classes that we could use in this project, like any remotely responsible person.

I would not be happy copying it's logic into this gem, I removed it in the first place because I use that particular functionality elsewhere on its own all over the place -- i.e. any ruby project I've got that's interacting with a camel cased API -- and when edge cases pop up I can fix and test them in one place and so on and so forth.

I'll go do that and let you know when it's done if you want to implement https://github.com/buermann/introspective_grape/pull/5 without those particular bits of this PR.

File: lib/introspective_grape/utils/json_expander.rb (hunk 1 above)
Comment (buermann) [reply lib/introspective_grape/utils/json_expander.rb:12]: Originally it was just `JSON.parse(object.to_json)`ing everything that responded to to_json (in my usage cases everything was a Grape::Entity so everything needed it), you were using the CamelJson formatter in your projects without Grape::Entity and reasonably wanted to skip that step.

**Comment by buermann** [comment]
camel_snake_keys no longer monkey patches Enumerable, the master branch of introspective_grape now uses the class library calls instead (JRuby doesn't support refinements).
